lingui.config.ts

```diff
@@ -1,8 +1,60 @@
+import { default as babelExtractor } from '@lingui/cli/api/extractors/babel'
+import { createHash } from 'crypto'
+import { mkdirSync, readFileSync, writeFileSync } from 'fs'
+import * as path from 'path'
+import * as pkgUp from 'pkg-up'
+
+/**
+ * A custom caching extractor for CI.
+ * Falls back to the babelExtractor in a non-CI (ie local) environment.
+ * Caches a file's latest extracted content's hash, and skips re-extracting if it is already present in the cache.
+ * In CI, re-extracting files takes over one minute, so this is a significant savings.
+ */
+const cachingExtractor: typeof babelExtractor = {
+  match(filename: string) {
+    return babelExtractor.match(filename)
+  },
+  extract(filename: string, code: string, ...options: unknown[]) {
+    if (!process.env.CI) return babelExtractor.extract(filename, code, ...options)
+
+    // This runs from node_modules/@lingui/conf, so we need to back out to the root.
+    const pkg = pkgUp.sync()
+    if (!pkg) throw new Error('No root found')
+    const root = path.dirname(pkg)
+
+    const filePath = path.join(root, filename)
+    const file = readFileSync(filePath)
+    const hash = createHash('sha256').update(file).digest('hex')
+
+    const cacheRoot = path.join(root, 'node_modules/.cache/lingui')
+    mkdirSync(cacheRoot, { recursive: true })
+    const cachePath = path.join(cacheRoot, filename.replace(/\//g, '-'))
+    if (!process.argv.includes('--clean')) {
+      try {
+        const cache = readFileSync(cachePath, 'utf8')
+        if (cache === hash) return
+      } catch (e) {
+        // It should not be considered an error if there is no cache file.
+      }
+    }
```

Comment:
async readFile might read better here

Reply:
I think this function still needs to return synchronously, though